libbpfgo.go

```diff
@@ -750,6 +754,90 @@ func doAttachKprobe(prog *BPFProg, kp string, isKretprobe bool) (*BPFLink, error
 	return bpfLink, nil
 }
 
+// AttachUprobe attaches the BPFProgram to entry of the symbol in the library or binary at 'path'
+// which can be relative or absolute. A pid can be provided to attach to, or -1 can be specified
+// to attach to all processes
+func (p *BPFProg) AttachUprobe(pid int, path string, symbol string) (*BPFLink, error) {
+	offset, err := symbolToOffset(path, symbol)
+	if err != nil {
+		return nil, err
+	}
+	absPath, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return doAttachUprobe(p, false, pid, absPath, offset)
+}
+
+// AttachURetprobe attaches the BPFProgram to exit of the symbol in the library or binary at 'path'
+// which can be relative or absolute. A pid can be provided to attach to, or -1 can be specified
+// to attach to all processes
+func (p *BPFProg) AttachURetprobe(pid int, path string, symbol string) (*BPFLink, error) {
+	offset, err := symbolToOffset(path, symbol)
+	if err != nil {
+		return nil, err
+	}
+
+	absPath, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return doAttachUprobe(p, true, pid, absPath, offset)
+}
+
+func symbolToOffset(path, symbol string) (uint32, error) {
+
+	f, err := elf.Open(path)
+	if err != nil {
+		return 0, fmt.Errorf("could not open elf file to resolve symbol offset: %v", err)
+	}
+
+	syms, err := f.Symbols()
+	if err != nil {
+		return 0, fmt.Errorf("could not open symbol section to resolve symbol offset: %v", err)
+	}
+
+	textSection := f.Section(".text")
```

Comment:
I'm not sure that only looking into the .text section is enough.
Although usually this will work, there are cases where the text section is named differently or there is more than one text section (with different names)

Reply:
I changed the AttachU(ret)probe API to accept an offset. I moved the ResolveOffset helper function to the helpers package. That function takes a symbol name and section, so leave it in our users hands 👍

Reply:
Example in selftest